PYIC-6156: Updated handleJourneyPage/Action session validation

AmritSidhu opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

- Seperate the session validation and page logic into a new funcition. This is now used in both the GET (handleJourneyPage) and POST (handleJourneyAction) handlers.
- Remove duplicated test
- Use correct page template ID's in test cases now that the POST (handleJourneyAction) function also checks if page is valid
- Add session.currentPage to test cases as the POST (handleJourneyAction) function also checks if session.currentPage is valid

### Why did it change

The GET validation is not the same as for POST. So people can/ can't get onto a page with different conditionality to whether they can post to it

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-6156](https://govukverify.atlassian.net/browse/PYIC-6156)

## Checklists

### Environment variables or secrets

<!-- Delete if changes DO include new environment variables or secrets -->
- [x] No environment variables or secrets were added or changed


[PYIC-6156]: https://govukverify.atlassian.net/browse/PYIC-6156?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

AmritSidhu committed:
PYIC-6156: Updated handleJourneyPage/Action session validation

- Seperate the session validation and page logic into a new funcition. This is now used in both the GET (handleJourneyPage) and POST (handleJourneyAction) handlers.
- Remove duplicated test
- Use correct page template ID's in test cases now that the POST (handleJourneyAction) function also checks if page is valid
- Add session.currentPage to test cases as the POST (handleJourneyAction) function also checks if session.currentPage is valid

DanCorderIPV commented on src/app/ipv/middleware.js:
```diff
@@ -330,6 +320,51 @@ async function renderAttemptRecoveryPage(req, res) {
+  // Check if the page is valid
+  if (!isValidIpvPage(pageId)) {
+    res.status(HTTP_STATUS_CODES.NOT_FOUND);
+    return res.render(getTemplatePath("errors", "page-not-found"));
```
We now have the `render404()` function

DanCorderIPV commented on src/app/ipv/middleware.js:
```diff
@@ -406,21 +423,13 @@ async function handleJourneyPage(req, res, next, pageErrorState = undefined) {
-    if (pagesNotUsingOAuthSessionId.includes(currentPageId)) {
-      checkForIpvSessionId(req, res);
-    } else {
-      checkForIpvAndOauthSessionId(req, res);
```
Is it safe to drop the check for the OAuth session ID?

AmritSidhu replied:
I believe the only scenario we have a `clientOauthSessionId` is for the page `pyi-timeout-recoverable` (the mobile cross-browser scenario). I moved the check to `validateSessionAndPage` function which checks specifically for that page as well as the ID

AmritSidhu committed:
PYIC-6156: Use new render404 function

blakeyp commented on src/app/ipv/middleware.js:
```diff
@@ -330,6 +320,50 @@ async function renderAttemptRecoveryPage(req, res) {
+  }
+
+  if (req.session.currentPage !== pageId) {
+    await handleUnexpectedPage(req, res, pageId);
```
Hmm can we just `return await` this rather than this function returning true/false?

blakeyp commented on src/app/ipv/middleware.js:
```diff
@@ -350,27 +384,8 @@ async function handleJourneyPage(req, res, next, pageErrorState = undefined) {
-      return res.render(getIpvPageTemplatePath(req.session.currentPage));
-    } else if (req.session.currentPage !== pageId) {
-      return await handleUnexpectedPage(req, res, pageId);
+    if (!(await validateSessionAndPage(req, res, pageId))) {
+      return;
```
and just call `await validateSessionAndPage(req, res, pageId)` here?

DanCorderIPV replied:
Won't we end up calling `res.render()` twice? Do we know if that's safe? A quick look at the docs didn't turn anything up.

AmritSidhu committed:
PYIC-6156: Update validateSessionAndPage

The function now returns the result of each conditional check (such as rendering a page) without returning true or false. This makes the function simpler and focused on handling the conditions rather than managing state.
Remove redundant return

AmritSidhu committed:
PYIC-6156: Add test cases for clientOauthSessionId and updateJourneyState

SonarCloud complaining about code coverage

blakeyp commented on src/app/ipv/middleware.test.js:
```diff
@@ -876,6 +939,76 @@ describe("journey middleware", () => {
+  context(
+    "handleJourneyAction: should render timeout-recoverable page if present clientOauthSessionId",
+    () => {
+      it("should render the technical unrecoverable page", async function () {
```
*timeout-recoverable

AmritSidhu committed:
PYIC-6156: Add return statements in handleJourneyPage and handleJourneyAction to prevent double res.render() calls
